src/rgw/rgw_file.cc

```diff
@@ -409,8 +409,15 @@ namespace rgw {
       goto unlock;
     }
 
-    /* forbid renaming open files (violates intent, for now) */
-    if (rgw_fh->is_open()) {
+    /* As NFS v3 is stateless, we need to make sure the 
+     * source object is closed.
+     */
+    if (rgw_fh->stateless_open()) {
+      rgw_fh->mtx.unlock(); /* !LOCKED */
```

Comment:
Looking at the code, it looks like to me the close, or rather the RGWFileHandle::write_finish what is called by the close will lock if the fh isn't locked. So the unlock and lock may be redundant. 

Though I should go test this just to be sure.

Reply:
we are changing the state, so it should be locked over the operation;  first, tho, if we reach l. 404 (i.e., we got a filehandle from lookup_fh) rgw_fh is locked;  if we want to hold the lock safely over the whole operation, we should look into passing a flags argument to RGWFileHandle::close(...) with the lock state?

Reply:
Just tested this. I could successfully move the object withoug the `rgw_fh->mtx.unlock();` dance. The rgw_fh object will check to see if it has a lock and if not grab it. the unlock then lock is redundant. 

So:
```
if (rgw_fh->stateless_open()) {
  rgw_fh->close();
} else if (rgw_fh->is_open()) {
```
Is enough.

Also note, I tested this on a rebase, because the line numbers are quite a bit out now. So I also suggest a rebase.

Reply:
as I read the linked version of RGWFileHandle::close(), I'd say that's suspect.  In the prologue, we have:```
  int RGWFileHandle::close()
  {
    lock_guard guard(mtx);
    int rc = write_finish(FLAG_LOCKED);
```
which is unconditionally taking this same mutex.  to deal with this, we should not only not drop the lock which rename() took, but rename should pass the (locked) file handle to close with an argument of FLAG_LOCKED.  (In order to do that, close needs to have a uint32_t flags argument that defaults to FLAG_NONE, e.g:
in class RGWFileHandle, rgw_file.h:
```
int close(uint32_t flags = FLAG_NONE);
```